controllers/apim/authpolicy_controller.go

```diff
@@ -0,0 +1,211 @@
+	}
+
+	for _, parentRef := range httpRoute.Spec.ParentRefs {
+		gwNamespace := httpRoute.Namespace // consider gateway local if namespace is not given
+		if parentRef.Namespace != nil {
+			gwNamespace = string(*parentRef.Namespace)
+		}
+		gwName := string(parentRef.Name)
+
+		// convert []Rule  to []*Rule
+		rulePtrSlice := []*secv1beta1types.Rule{}
+		for idx := range ap.Spec.Rules {
+			rulePtrSlice = append(rulePtrSlice, &ap.Spec.Rules[idx])
+		}
+
+		actionInt := secv1beta1types.AuthorizationPolicy_Action_value[string(ap.Spec.Action)]
+
+		authPolicy := secv1beta1resources.AuthorizationPolicy{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      getAuthPolicyName(gwName, httpRoute.Name),
+				Namespace: gwNamespace,
+			},
+			Spec: secv1beta1types.AuthorizationPolicy{
+				Action: secv1beta1types.AuthorizationPolicy_Action(actionInt),
+				Rules:  rulePtrSlice,
+				ActionDetail: &secv1beta1types.AuthorizationPolicy_Provider{
+					Provider: &ap.Spec.Provider,
+				},
+				Selector: &v1beta1.WorkloadSelector{
+					MatchLabels: map[string]string{}, // TODO(rahulanand16nov): fetch from gateway
+				},
+			},
+		}
+
+		err := r.ReconcileResource(ctx, &secv1beta1resources.AuthorizationPolicy{}, &authPolicy, alwaysUpdateAuthPolicy)
+		if err != nil && !apierrors.IsAlreadyExists(err) {
+			logger.Error(err, "ReconcileResource failed to create/update AuthorizationPolicy resource")
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *AuthPolicyReconciler) removeIstioAuthPolicy(ctx context.Context, ap *apimv1alpha1.AuthPolicy) error {
+	logger := logr.FromContext(ctx)
+	logger.Info("Removing Istio's AuthorizationPolicy")
+
+	httpRoute, err := r.fetchHTTPRoute(ctx, ap)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("referenced HTTPRoute not found")
+			return nil
+		}
+		return err
+	}
+
+	for _, parentRef := range httpRoute.Spec.ParentRefs {
+		gwNamespace := httpRoute.Namespace
+		if parentRef.Namespace != nil {
+			gwNamespace = string(*parentRef.Namespace)
+		}
+		gwName := string(parentRef.Name)
+
+		authPolicyKey := client.ObjectKey{
+			Namespace: gwNamespace,
+			Name:      getAuthPolicyName(gwName, httpRoute.Name),
+		}
+
+		istioAuthPolicy := &secv1beta1resources.AuthorizationPolicy{}
+		if err := r.GetResource(ctx, authPolicyKey, istioAuthPolicy); err != nil {
+			logger.Error(err, "failed to fetch Istio's AuthorizationPolicy")
+			return err
+		}
+
+		if err := r.DeleteResource(ctx, istioAuthPolicy); err != nil {
+			logger.Error(err, "failed to delete Istio's AuthorizationPolicy")
+			return err
+		}
+	}
+
+	logger.Info("removed Istio's AuthorizationPolicy")
+	return nil
+}
+
+// fetchHTTPRoute fetches the HTTPRoute described in targetRef *within* AuthPolicy's namespace.
+func (r *AuthPolicyReconciler) fetchHTTPRoute(ctx context.Context, ap *apimv1alpha1.AuthPolicy) (*gatewayapiv1alpha2.HTTPRoute, error) {
+	logger := logr.FromContext(ctx)
+	key := client.ObjectKey{
+		Name:      string(ap.Spec.TargetRef.Name),
```

Comment:
I don't see any check that this TargetRef is actually to a HTTPRoute? Am I missing something? HTH

Reply:
Hey Mark, It's still in the development stages, but right now, we allow (assume) it to only target an HTTPRoute, but with the next PR, it should be able to target both `Gateway` and `HTTPRoute`.

Reply:
You can find the validation checks in #146

Reply:
Thanks, I just found the Validate() methods while looking at the RateLimitPolicy example!